Mathlib/Geometry/RingedSpace/OpenImmersion.lean

```diff
@@ -865,6 +865,95 @@ theorem of_stalk_iso {X Y : SheafedSpace C} (f : X ⟶ Y) (hf : OpenEmbedding f.
 
 end OfStalkIso
 
+section
+
+variable {X Y : SheafedSpace C} {f : X ⟶ Y} (H : IsOpenImmersion f)
+
+/-- The functor `opens X ⥤ opens Y` associated with an open immersion `f : X ⟶ Y`. -/
+abbrev openFunctor :=
+  H.base_open.isOpenMap.functor
+
+/-- An open immersion `f : X ⟶ Y` induces an isomorphism `X ≅ Y|_{f(X)}`. -/
+@[simps! hom_c_app]
+noncomputable def isoRestrict : X ≅ Y.restrict H.base_open :=
+  SheafedSpace.isoMk <| PresheafedSpace.IsOpenImmersion.isoRestrict H
+
+@[simp]
+theorem isoRestrict_hom_ofRestrict : H.isoRestrict.hom ≫ Y.ofRestrict _ = f :=
+  PresheafedSpace.IsOpenImmersion.isoRestrict_hom_ofRestrict H
+
+@[simp]
+theorem isoRestrict_inv_ofRestrict : H.isoRestrict.inv ≫ f = Y.ofRestrict _ :=
+  PresheafedSpace.IsOpenImmersion.isoRestrict_inv_ofRestrict H
+
+/-- For an open immersion `f : X ⟶ Y` and an open set `U ⊆ X`, we have the map `X(U) ⟶ Y(U)`. -/
+noncomputable def invApp (U : Opens X) :
+    X.presheaf.obj (op U) ⟶ Y.presheaf.obj (op (H.openFunctor.obj U)) :=
+  PresheafedSpace.IsOpenImmersion.invApp H U
+
+@[simp, reassoc]
+theorem inv_naturality {U V : (Opens X)ᵒᵖ} (i : U ⟶ V) :
+    X.presheaf.map i ≫ H.invApp (unop V) =
+      H.invApp (unop U) ≫ Y.presheaf.map (H.openFunctor.op.map i) :=
+  PresheafedSpace.IsOpenImmersion.inv_naturality H i
+
+instance (U : Opens X) : IsIso (H.invApp U) := by delta invApp; infer_instance
+
+theorem inv_invApp (U : Opens X) :
+    inv (H.invApp U) =
+      f.c.app (op (H.openFunctor.obj U)) ≫
+        X.presheaf.map (eqToHom (by simp [Opens.map, Set.preimage_image_eq _ H.base_open.inj])) :=
+  PresheafedSpace.IsOpenImmersion.inv_invApp H U
+
+@[simp, reassoc, elementwise]
```

Comment:
I am not sure how to get all the expected attributes at once. You may try to add a line `attribute [simp] invApp_app_assoc`.

Reply:
I am not sure if I understand the comment correctly, I did the following:
```
@[reassoc (attr := simp)] 
lemma foo := ….

attribute [elementwise] foo
```